src/full/Agda/Interaction/Highlighting/HTML.hs

```diff
@@ -205,17 +206,28 @@ tokenStream contents info =
 code :: Bool          -- ^ Whether to generate non-code contents as-is
      -> [(Int, String, Aspects)]
      -> Html
-code preserveLiterate = mconcat . snd . List.mapAccumL mkHtml True
+code asIs = mconcat . if asIs
+  then map mkMd . chunksOf 2 . splitWhen ((== Just Markup) . aspect . trd)
+  else map mkHtml
   where
-  mkHtml :: Bool -> (Int, String, Aspects) -> (Bool, Html)
-  mkHtml pre (pos, s, mi) = case (preserveLiterate, aspect mi) of
-    (True, Just Background) -> (pre, preEscapedToHtml s)
-    (True, Just Markup)     -> (not pre, preEscapedToHtml s')
-      where s' :: String
-            s' = if pre then "<pre class=\"agda-code\">" else "</pre>"
+  trd (_, _, a) = a
+
+  mkHtml :: (Int, String, Aspects) -> Html
+  mkHtml (pos, s, mi) =
     -- Andreas, 2017-06-16, issue #2605:
     -- Do not create anchors for whitespace.
-    _ -> (pre, applyUnless (mi == mempty) (annotate pos mi) $ toHtml s)
+    applyUnless (mi == mempty) (annotate pos mi) $ toHtml s
+
+  mkMd :: [[(Int, String, Aspects)]] -> Html
+  mkMd = go
+    where
+      work a@(pos, s, mi) = case aspect mi of
+        Just Background -> preEscapedToHtml s
+        Just Markup     -> __IMPOSSIBLE__
+        _               -> mkHtml a
+      go [a, b] = mconcat [mconcat $ work <$> a, pre ! class_ "agda-code" $ mconcat $ work <$> b]
```

Comment:
It still works like this:

![image](https://user-images.githubusercontent.com/16398479/47584810-a2062b80-d929-11e8-9037-f93bf84484f0.png)

Very strange